Choreo Connect 1.0.0 - Helm QSG and Production deployment guide (#5214)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/production-deployment-guideline.md

# Production Deployment Guideline

Given below are the common guidelines for deploying Choreo Connect in a **production environment** using [Helm deployment option](http://localhost:8000/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/deploy/cc-deploy-overview/).

## Configuring Keystore

By default, Choreo Connect ships with default keystore certificates, and it is recommended to replace them with Self-Signed or CA signed certificates.

> ### Setep 1: Generate Self-Signed certs in PEM format

Reviewer: ```suggestion
### Step 1: Generate self-signed certs in PEM format
```

    -keyout <KEY_FILE_PATH> \
    -out <CERT_FILE_PATH>
```

```shell tab='Adapter Sample'
openssl req -x509 -sha256 -nodes -days 3650 -newkey rsa:2048 \
    -subj "/C=US/ST=CA/L=Mountain View/O=WSO2, Inc./OU=Choreo Connect/CN=yourdomain.com" \
    -extensions SAN \
    -config <(cat /etc/ssl/openssl.cnf \
        <(printf "\n[SAN]\nsubjectAltName=DNS:adapter")) \
    -keyout certs/adapter/tls.key \
    -out certs/adapter/tls.crt
```

```shell tab='Enforcer Sample'
openssl req -x509 -sha256 -nodes -days 3650 -newkey rsa:2048 \
    -subj "/C=US/ST=CA/L=Mountain View/O=WSO2, Inc./OU=Choreo Connect/CN=yourdomain.com" \
    -extensions SAN \
    -config <(cat /etc/ssl/openssl.cnf \
        <(printf "\n[SAN]\nsubjectAltName=DNS:enforcer")) \
    -keyout certs/enforcer/tls.key \
    -out certs/enforcer/tls.crt
```

```shell tab='Router Sample'
openssl req -x509 -sha256 -nodes -days 3650 -newkey rsa:2048 \
    -subj "/C=US/ST=CA/L=Mountain View/O=WSO2, Inc./OU=Choreo Connect/CN=yourdomain.com" \
    -extensions SAN \
    -config <(cat /etc/ssl/openssl.cnf \
        <(printf "\n[SAN]\nsubjectAltName=DNS:router")) \
    -keyout certs/router/tls.key \
    -out certs/router/tls.crt
```

> By default, SSL Hostnames for adapter and enforcer is `adapter` and `enforcer`, in case if you want to have your own SAN names in the certificates, update the following values in helm chart with the SAN names in the certificates.

Reviewer: ```suggestion
By default, SSL Hostnames for adapter and enforcer is `adapter` and `enforcer`, in case if you want to have your own SAN names in the certificates, update the following values in the Helm Chart with the SAN names in the certificates.
```

```

## Configure Ingress Secrets

By default, the helm chart installation of Choreo Connect creates default certificates for ingress resources.
Providing TLS Secret name in the following values will replace those default certs with provided certs.

!!! Note
    Following certs should be [TLS secrets](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets)

```shell tab='Adapter Ingress'
wso2.deployment.adapter.ingress.tlsSecretName=<TLS_CERT_SECRET_IN_THE_SAME_NAMESPACE>
```

```shell tab='Gateway Ingress'
wso2.deployment.gatewayRuntime.router.ingress.tlsSecretName=<TLS_CERT_SECRET_IN_THE_SAME_NAMESPACE>
```

## Disable Test Token Issuer

Choreo Connect provides the ability to generate a JWT as a test key to access the APIs. It is recommended to disable the `/testkey` endpoint in the Production environments.
Refer the document [Generate a Test JWT]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/generate-a-test-jwt/) to read more about Test JWT.

Set the following value to `false` when installing the helm chart or set it in the values.yaml file.

```shell
wso2.deployment.gatewayRuntime.enforcer.security.testTokenIssuer.enabled=false
```

> ## Create Custom Docker Image

Author: Make Adapter custom docker image

Author: Added separate sections for Standalone and Control Plane modes with https://github.com/wso2/docs-apim/pull/5214/commits/9519594eb8494b506a1b2318846d599edf1cf7f6

You can create your own docker image using Choreo Connect docker images as the base image.

For example lets say you want to replace the CA certificates of Choreo Connect Router with your own set of certificates.
Create a Dockerfile as follows.

```dockerfile tab='Format'
FROM wso2/choreo-connect-router:1.0.0
<YOUR_DOCKER_COMMANDS>
```

```dockerfile tab='Sample'
FROM wso2/choreo-connect-router:1.0.0
COPY my-ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
```

> make sure the files you are adding or coping (with commands `ADD` and `COPY`) exists in the context path.

Reviewer: ```suggestion
Make sure the files you are adding or copying (with commands `ADD` and `COPY`) exist in the context path.
```
